// ngx-datetime-picker/esm5/services/date.service.js
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
var DateService = /** @class */ (function () {
    function DateService() {
        this.months = ['January', 'February', 'March', 'April', 'May', 'June', 'July', 'August', 'September', 'October', 'November', 'December'];
    }
    DateService.prototype.addLeadingZero = function (value) {
        if (value < 10) {
            return "0" + value.toString();
        }
        return value.toString();
    };
    DateService.prototype.formatMobileYYYYMMDD = function (date) {
        if (!date || typeof date == 'string') {
            return '';
        }
        return date.getFullYear() + "-" + this.addLeadingZero(date.getMonth() + 1) + "-" + this.addLeadingZero(date.getDate());
    };
    DateService.prototype.formatMobileYYYYMMDDTHHMM = function (date) {
        if (!date || typeof date == 'string') {
            return '';
        }
        return this.formatMobileYYYYMMDD(date) + "T" + this.addLeadingZero(date.getHours()) + ":" + this.addLeadingZero(date.getMinutes());
    };
    DateService.prototype.formatMMDDYYYY = function (date) {
        if (!date || typeof date == 'string') {
            return '';
        }
        return (date.getMonth() + 1) + "/" + date.getDate() + "/" + date.getFullYear();
    };
    DateService.prototype.formatMMDDYYYY_HHMM_AMPM = function (date) {
        if (!date || typeof date == 'string') {
            return '';
        }
        var hours = date.getHours();
        var minutes = date.getMinutes();
        return this.formatMMDDYYYY(date) + " " + this.formatHHMM_AMPM(hours, minutes);
    };
    DateService.prototype.formatMMDDYYYY_HHMM = function (date) {
        if (!date || typeof date == 'string') {
            return '';
        }
        var hours = date.getHours();
        var minutes = date.getMinutes();
        return this.formatMMDDYYYY(date) + " " + this.formatHHMM(hours, minutes);
    };
    DateService.prototype.formatHHMM_AMPM = function (hour, minute, clock) {
        if (hour == null || minute == null) {
            return '';
        }
        if (typeof clock === 'undefined') {
            clock = hour >= 12 ? 'pm' : 'am';
        }
        if (hour > 12) {
            hour = hour - 12;
        }
        else if (hour === 0) {
            hour = 12;
        }
        var formattedMinute = (!minute ? '00' : (minute <= 9 ? "0" + minute : minute));
        return hour + ":" + formattedMinute + " " + clock;
    };
    DateService.prototype.formatHHMM = function (hour, minute) {
        if (hour == null || minute == null) {
            return '';
        }
        var formattedHour = (!hour ? '00' : (hour <= 9 ? "0" + hour : hour));
        var formattedMinute = (!minute ? '00' : (minute <= 9 ? "0" + minute : minute));
        return formattedHour + ":" + formattedMinute;
    };
    DateService.prototype.getCurrentMonthDays = function (month, year) {
        var dayOfTheMonth = new Date(year, month - 1, 1);
        var nextMonth = new Date(year, month - 1, 1);
        var returnedDays = [];
        nextMonth.setMonth(nextMonth.getMonth() + 1);
        while (dayOfTheMonth.getMonth() != nextMonth.getMonth()) {
            var dayToAdd = {
                day: dayOfTheMonth.getDate(),
                dayOfTheWeek: dayOfTheMonth.getDay(),
                month: dayOfTheMonth.getMonth() + 1,
                date: new Date((dayOfTheMonth.getMonth() + 1) + '/' + dayOfTheMonth.getDate() + '/' + dayOfTheMonth.getFullYear())
            };
            returnedDays.push(dayToAdd);
            dayOfTheMonth.setDate(dayOfTheMonth.getDate() + 1);
        }
        return returnedDays;
    };
    DateService.prototype.getDateList = function (Month, Year) {
        return tslib_1.__spread(this.getPreviousMonthDays(Month, Year), this.getCurrentMonthDays(Month, Year), this.getNextMonthDays(Month, Year));
    };
    DateService.prototype.getPreviousMonthDays = function (Month, Year) {
        var day = new Date(Month + '/1/' + Year);
        var returnedDays = [];
        var dayOfTheWeek = day.getDay();
        while (dayOfTheWeek != 0) {
            day.setDate(day.getDate() - 1);
            returnedDays = tslib_1.__spread([{
                    day: day.getDate(),
                    dayOfTheWeek: day.getDay(),
                    month: day.getMonth() + 1,
                    date: new Date((day.getMonth() + 1) + '/' + day.getDate() + '/' + day.getFullYear())
                }], returnedDays);
            dayOfTheWeek = day.getDay();
        }
        return returnedDays;
    };
    DateService.prototype.getNextMonthDays = function (Month, Year) {
        var day = new Date(Month + '/1/' + Year);
        day.setMonth(day.getMonth() + 1);
        day.setDate(day.getDate() - 1);
        var returnedDays = [];
        var dayOfTheWeek = day.getDay();
        while (dayOfTheWeek != 6) {
            day.setDate(day.getDate() + 1);
            returnedDays = tslib_1.__spread(returnedDays, [{
                    day: day.getDate(),
                    dayOfTheWeek: day.getDay(),
                    month: day.getMonth() + 1,
                    date: new Date((day.getMonth() + 1) + '/' + day.getDate() + '/' + day.getFullYear())
                }]);
            dayOfTheWeek = day.getDay();
        }
        return returnedDays;
    };
    DateService.prototype.getMonths = function () {
        return this.months;
    };
    DateService.prototype.getMonthText = function (date) {
        if (date == null) {
            date = new Date();
        }
        return this.months[date.getMonth()];
    };
    DateService.prototype.getAvailableYears = function () {
        var currentYear = new Date().getFullYear();
        var startYear = currentYear - 80;
        var returnYears = [];
        while (startYear <= (currentYear + 5)) {
            returnYears.push(startYear);
            startYear = startYear + 1;
        }
        return returnYears;
    };
    DateService.prototype.canSelectYear = function (year, min, max) {
        if (!min && !max) {
            return true;
        }
        if (min && year < new Date(min).getFullYear()) {
            return false;
        }
        if (max && year > new Date(max).getFullYear()) {
            return false;
        }
        return true;
    };
    DateService.prototype.canSelectMonth = function (month, year, min, max) {
        if (!min && !max) {
            return true;
        }
        if ((min && year === new Date(min).getFullYear())
            || (max && year === new Date(max).getFullYear())) {
            if (min && year === new Date(min).getFullYear() && month < new Date(min).getMonth()) {
                return false;
            }
            if (max && year === new Date(max).getFullYear() && month > new Date(max).getMonth()) {
                return false;
            }
            return true;
        }
        else {
            return this.canSelectYear(year, min, max);
        }
    };
    DateService.prototype.canSelectDay = function (day, month, year, min, max) {
        if (!min && !max) {
            return true;
        }
        if ((min && (year === new Date(min).getFullYear() && month === new Date(min).getMonth()))
            || (max && (year === new Date(max).getFullYear() && month === new Date(max).getMonth()))) {
            if (min && year === new Date(min).getFullYear() && month === new Date(min).getMonth() && day < new Date(min).getDate()) {
                return false;
            }
            if (max && year === new Date(max).getFullYear() && month === new Date(max).getMonth() && day > new Date(max).getDate()) {
                return false;
            }
            return true;
        }
        else {
            return this.canSelectMonth(month, year, min, max);
        }
    };
    DateService = tslib_1.__decorate([
        Injectable()
    ], DateService);
    return DateService;
}());
export { DateService };
