// ngx-datetime-picker/esm5/components/datePicker/datePicker.component.js
import * as tslib_1 from "tslib";
import { Component, OnInit, Input, Output, EventEmitter, ViewEncapsulation, ElementRef, forwardRef, ViewChild, HostListener } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { IsMobileService } from '../../services/isMobile.service';
import { DateService } from '../../services/date.service';
import { StyleObject } from '../../models/styleObject.model';
import { Renderer } from '../../services/renderer.service';
var DatePickerComponent = /** @class */ (function () {
    function DatePickerComponent(isMobileService, dateService, eRef, renderer) {
        this.isMobileService = isMobileService;
        this.dateService = dateService;
        this.eRef = eRef;
        this.renderer = renderer;
        this.disableInput = false;
        this.disableButton = false;
        this.disablePicker = false;
        this.doNotCloseOnDateSet = false;
        this.styles = new StyleObject();
        this.selectedDateChange = new EventEmitter();
        this.pickerVisible = false;
        this.isMobile = isMobileService.isMobile;
        this.placeholder = this.placeholder || '';
    }
    DatePickerComponent_1 = DatePickerComponent;
    DatePickerComponent.prototype.offClick = function (event) {
        if (!this.eRef.nativeElement.contains(event.target)) {
            this.pickerVisible = false;
        }
    };
    Object.defineProperty(DatePickerComponent.prototype, "formattedDate", {
        get: function () {
            return this.dateService.formatMMDDYYYY(this.selectedDate);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DatePickerComponent.prototype, "mobileFormattedDate", {
        get: function () {
            return this.dateService.formatMobileYYYYMMDD(this.selectedDate);
        },
        enumerable: true,
        configurable: true
    });
    DatePickerComponent.prototype.writeValue = function (value) {
        this.selectedDate = value;
    };
    DatePickerComponent.prototype.registerOnChange = function (handler) {
        this.selectedDateChange.subscribe(handler);
    };
    DatePickerComponent.prototype.registerOnTouched = function () { };
    // for use with the native html5 element. only emit's new valid dates.
    DatePickerComponent.prototype.setDate = function (date) {
        this.invalid = !Date.parse(date + " 00:00:00");
        if (!this.invalid) {
            // set the time to zero so that values emitted on mobile are the same as on desktop
            this.selectedDate = new Date(date + " 00:00:00");
            this.selectedDateChange.emit(this.selectedDate);
        }
    };
    DatePickerComponent.prototype.ngOnInit = function () {
        if (typeof this.selectedDate == 'string') {
            this.selectedDate = new Date(this.selectedDate);
        }
    };
    DatePickerComponent.prototype.newDatePicked = function (date) {
        this.invalid = false;
        this.selectedDateChange.emit(date);
        this.selectedDate = date;
    };
    DatePickerComponent.prototype.setPickerVisible = function (close) {
        this.pickerVisible = close;
    };
    DatePickerComponent.prototype.focus = function () {
        this.renderer.invokeElementMethod(this.input.nativeElement, 'focus');
    };
    var DatePickerComponent_1;
    DatePickerComponent.ctorParameters = function () { return [
        { type: IsMobileService },
        { type: DateService },
        { type: ElementRef },
        { type: Renderer }
    ]; };
    tslib_1.__decorate([
        Input()
    ], DatePickerComponent.prototype, "selectedDate", void 0);
    tslib_1.__decorate([
        Input()
    ], DatePickerComponent.prototype, "min", void 0);
    tslib_1.__decorate([
        Input()
    ], DatePickerComponent.prototype, "max", void 0);
    tslib_1.__decorate([
        Input()
    ], DatePickerComponent.prototype, "placeholder", void 0);
    tslib_1.__decorate([
        Input()
    ], DatePickerComponent.prototype, "inputTabIndex", void 0);
    tslib_1.__decorate([
        Input()
    ], DatePickerComponent.prototype, "disableInput", void 0);
    tslib_1.__decorate([
        Input()
    ], DatePickerComponent.prototype, "disableButton", void 0);
    tslib_1.__decorate([
        Input()
    ], DatePickerComponent.prototype, "disablePicker", void 0);
    tslib_1.__decorate([
        Input()
    ], DatePickerComponent.prototype, "doNotCloseOnDateSet", void 0);
    tslib_1.__decorate([
        Input()
    ], DatePickerComponent.prototype, "styles", void 0);
    tslib_1.__decorate([
        Output()
    ], DatePickerComponent.prototype, "selectedDateChange", void 0);
    tslib_1.__decorate([
        ViewChild('input')
    ], DatePickerComponent.prototype, "input", void 0);
    tslib_1.__decorate([
        HostListener('document:click', ['$event'])
    ], DatePickerComponent.prototype, "offClick", null);
    DatePickerComponent = DatePickerComponent_1 = tslib_1.__decorate([
        Component({
            selector: 'ngx-date-picker',
            template: "<div [ngSwitch]=\"isMobile\"\n     [class.invalid]=\"invalid\">\n\t<div *ngSwitchCase=\"true\">\n\t\t<input type=\"date\" #input [disabled]=\"disableInput || disablePicker\" [placeholder]=\"placeholder\" [value]=\"mobileFormattedDate\" (change)=\"setDate($event.target.value)\" [tabindex]=\"inputTabIndex\" [min]=\"min\" [max]=\"max\" />\n\t</div>\n\t<div *ngSwitchDefault>\n\t\t<div class=\"ngx-picker\">\n\t\t\t<input type=\"text\"\n                   #input\n                   [ngStyle]=\"styles.input\"\n                   [disabled]=\"disableInput || disablePicker\"\n                   [placeholder]=\"placeholder\"\n                   (focusin)=\"setPickerVisible(true)\"\n                   [value]=\"formattedDate\"\n                   (change)=\"setDate($event.target.value)\"\n                   [tabindex]=\"inputTabIndex\"\n\t\t\t/>\n\n\t\t\t<button type=\"button\"\n                    [ngStyle]=\"styles.button\"\n                    [disabled]=\"disableButton || disablePicker\"\n                    (focusin)=\"setPickerVisible(true)\">\n\t\t\t\t<svg xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" version=\"1.1\"  width=\"24\" height=\"24\" viewBox=\"0 0 24 24\">\n\t\t\t\t\t<path fill=\"#000000\" d=\"M19,19H5V8H19M16,1V3H8V1H6V3H5C3.89,3 3,3.89 3,5V19A2,2 0 0,0 5,21H19A2,2 0 0,0 21,19V5C21,3.89 20.1,3 19,3H18V1M17,12H12V17H17V12Z\" />\n\t\t\t\t</svg>\n\t\t\t</button>\n\n\t\t\t<ngx-date *ngIf=\"pickerVisible\"\n                      [ngStyle]=\"styles.date\"\n                      [doNotCloseOnDateSet]=\"doNotCloseOnDateSet\"\n                      (closeDatePicker)=\"setPickerVisible($event)\"\n                      (selectedDateChange)=\"newDatePicked($event)\"\n                      [selectedDate]=\"selectedDate\"\n                      [min]=\"min\"\n                      [max]=\"max\">\n\t\t\t</ngx-date>\n\t\t</div>\n\t</div>\n</div>\n",
            encapsulation: ViewEncapsulation.None,
            providers: [
                {
                    provide: NG_VALUE_ACCESSOR,
                    useExisting: forwardRef(function () { return DatePickerComponent_1; }),
                    multi: true,
                },
            ]
        })
    ], DatePickerComponent);
    return DatePickerComponent;
}());
export { DatePickerComponent };
