// ngx-datetime-picker/esm2015/index.js
import * as tslib_1 from "tslib";
var DateTimePickerModule_1;
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { DateComponent } from './components/date/date.component';
import { DatePickerComponent } from './components/datePicker/datePicker.component';
import { TimeComponent } from './components/time/time.component';
import { DateTimePickerComponent } from './components/dateTimePicker/dateTimePicker.component';
import { TimePickerComponent } from './components/timePicker/timePicker.component';
export * from './components/datePicker/datePicker.component';
export * from './components/dateTimePicker/dateTimePicker.component';
export * from './components/timePicker/timePicker.component';
export * from './services/date.service';
import { DateService } from './services/date.service';
import { IsMobileService } from './services/isMobile.service';
import { Renderer } from './services/renderer.service';
let DateTimePickerModule = DateTimePickerModule_1 = class DateTimePickerModule {
    static forRoot() {
        return {
            ngModule: DateTimePickerModule_1,
            providers: []
        };
    }
};
DateTimePickerModule = DateTimePickerModule_1 = tslib_1.__decorate([
    NgModule({
        declarations: [
            TimeComponent,
            DateComponent,
            DatePickerComponent,
            DateTimePickerComponent,
            TimePickerComponent
        ],
        exports: [
            DatePickerComponent,
            DateTimePickerComponent,
            TimePickerComponent
        ],
        imports: [
            CommonModule
        ],
        providers: [
            DateService,
            IsMobileService,
            Renderer
        ],
    })
], DateTimePickerModule);
export { DateTimePickerModule };
