// ngx-datetime-picker/esm5/components/timePicker/timePicker.component.js
import * as tslib_1 from "tslib";
import { Component, ViewEncapsulation, EventEmitter, Input, Output, forwardRef, ElementRef, HostListener } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { IsMobileService } from '../../services/isMobile.service';
import { DateService } from '../../services/date.service';
import { StyleObject } from '../../models/styleObject.model';
var TimePickerComponent = /** @class */ (function () {
    function TimePickerComponent(isMobileService, dateService, eRef) {
        this.isMobileService = isMobileService;
        this.dateService = dateService;
        this.eRef = eRef;
        this.disableInput = false;
        this.disableButton = false;
        this.disablePicker = false;
        this.doNotCloseOnDateSet = false;
        this.styles = new StyleObject();
        this.use24HourClock = false;
        this.pickerVisible = false;
        this.selectedTimeChange = new EventEmitter();
        this.isMobile = isMobileService.isMobile;
        this.placeholder = this.placeholder || '';
    }
    TimePickerComponent_1 = TimePickerComponent;
    TimePickerComponent.prototype.offClick = function (event) {
        if (!this.eRef.nativeElement.contains(event.target)) {
            this.pickerVisible = false;
        }
    };
    Object.defineProperty(TimePickerComponent.prototype, "formattedTime", {
        get: function () {
            if (this.selectedTime == null) {
                return '';
            }
            this.selectedHour = parseInt(this.selectedTime.split(':')[0]);
            this.selectedMinute = parseInt(this.selectedTime.split(':')[1]);
            if (this.use24HourClock) {
                return this.dateService.formatHHMM(this.selectedHour, this.selectedMinute);
            }
            return this.dateService.formatHHMM_AMPM(this.selectedHour, this.selectedMinute);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TimePickerComponent.prototype, "mobileFormattedTime", {
        get: function () {
            if (this.selectedTime == null) {
                return '';
            }
            this.selectedHour = parseInt(this.selectedTime.split(':')[0]);
            this.selectedMinute = parseInt(this.selectedTime.split(':')[1]);
            return (this.selectedHour < 10 ? '0' + this.selectedHour : this.selectedHour) + ":" + (this.selectedMinute < 10 ? '0' + this.selectedMinute : this.selectedMinute);
        },
        set: function (value) {
            var hour = value.split(':')[0];
            var minute = value.split(':')[1];
            if (parseInt(hour)) {
                this.selectedHour = parseInt(hour);
            }
            else {
                this.selectedHour = 0;
            }
            if (parseInt(minute)) {
                this.selectedMinute = parseInt(minute);
            }
            else {
                this.selectedMinute = 0;
            }
            this.selectedTime = hour + ":" + minute + " " + (parseInt(hour) <= 11 ? 'am' : 'pm');
        },
        enumerable: true,
        configurable: true
    });
    TimePickerComponent.prototype.writeValue = function (value) {
        this.selectedTime = value;
    };
    TimePickerComponent.prototype.registerOnChange = function (handler) {
        this.selectedTimeChange.subscribe(handler);
    };
    TimePickerComponent.prototype.registerOnTouched = function () {
    };
    TimePickerComponent.prototype.setMobileFormattedTime = function (time) {
        this.selectedTimeChange.emit(time);
        this.selectedTime = time;
    };
    TimePickerComponent.prototype.setFormattedTime = function (formattedTime) {
        this.selectedTime = formattedTime;
        this.selectedTimeChange.emit(formattedTime);
    };
    TimePickerComponent.prototype.setHourNow = function (hour) {
        var clock = hour <= 11 ? 'am' : 'pm';
        if (this.selectedTime == null || this.selectedTime === '') {
            this.selectedTime = hour + ":00 " + clock;
        }
        else {
            var prevMinute = parseInt(this.selectedTime.split(':')[1]);
            this.selectedTime = hour + ":" + prevMinute + " " + clock;
        }
        this.selectedTimeChange.emit(this.selectedTime);
    };
    TimePickerComponent.prototype.setMinuteNow = function (minute) {
        if (this.selectedTime == null || this.selectedTime === '') {
            this.selectedTime = "12:" + minute + " pm";
        }
        else {
            var prevHour = parseInt(this.selectedTime.split(':')[0]);
            this.selectedTime = prevHour + ":" + minute + " " + (prevHour <= 11 ? 'am' : 'pm');
        }
        this.selectedTimeChange.emit(this.selectedTime);
    };
    TimePickerComponent.prototype.setPickerVisible = function (close) {
        this.pickerVisible = close;
    };
    var TimePickerComponent_1;
    TimePickerComponent.ctorParameters = function () { return [
        { type: IsMobileService },
        { type: DateService },
        { type: ElementRef }
    ]; };
    tslib_1.__decorate([
        Input()
    ], TimePickerComponent.prototype, "selectedTime", void 0);
    tslib_1.__decorate([
        Input()
    ], TimePickerComponent.prototype, "placeholder", void 0);
    tslib_1.__decorate([
        Input()
    ], TimePickerComponent.prototype, "disableInput", void 0);
    tslib_1.__decorate([
        Input()
    ], TimePickerComponent.prototype, "disableButton", void 0);
    tslib_1.__decorate([
        Input()
    ], TimePickerComponent.prototype, "disablePicker", void 0);
    tslib_1.__decorate([
        Input()
    ], TimePickerComponent.prototype, "doNotCloseOnDateSet", void 0);
    tslib_1.__decorate([
        Input()
    ], TimePickerComponent.prototype, "styles", void 0);
    tslib_1.__decorate([
        Input()
    ], TimePickerComponent.prototype, "use24HourClock", void 0);
    tslib_1.__decorate([
        Output()
    ], TimePickerComponent.prototype, "selectedTimeChange", void 0);
    tslib_1.__decorate([
        HostListener('document:click', ['$event'])
    ], TimePickerComponent.prototype, "offClick", null);
    TimePickerComponent = TimePickerComponent_1 = tslib_1.__decorate([
        Component({
            selector: 'ngx-time-picker',
            template: "<div [ngSwitch]=\"isMobile\">\n\t<div *ngSwitchCase=\"true\">\n\t\t<input type=\"time\"\n               [disabled]=\"disableInput || disablePicker\"\n               [placeholder]=\"placeholder\"\n               [value]=\"mobileFormattedTime\"\n               (change)=\"setMobileFormattedTime($event.target.value)\"\n\t\t/>\n\t</div>\n\n\t<div *ngSwitchDefault>\n\t\t<div class=\"ngx-picker\">\n\t\t\t<input type=\"text\"\n                   [ngStyle]=\"styles.input\"\n                   [disabled]=\"disableInput || disablePicker\"\n                   [placeholder]=\"placeholder\"\n                   [value]=\"formattedTime\"\n                   (change)=\"setFormattedTime($event.target.value)\"\n                   (focusin)=\"setPickerVisible(true)\"/>\n\n\t\t\t<button type=\"button\"\n                    [ngStyle]=\"styles.button\"\n                    [disabled]=\"disableButton || disablePicker\"\n                    (focusin)=\"setPickerVisible(true)\">\n\t\t\t\t<svg xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" version=\"1.1\"  width=\"24\" height=\"24\" viewBox=\"0 0 24 24\">\n\t\t\t\t\t<path fill=\"#000000\" d=\"M12,20A8,8 0 0,0 20,12A8,8 0 0,0 12,4A8,8 0 0,0 4,12A8,8 0 0,0 12,20M12,2A10,10 0 0,1 22,12A10,10 0 0,1 12,22C6.47,22 2,17.5 2,12A10,10 0 0,1 12,2M12.5,7V12.25L17,14.92L16.25,16.15L11,13V7H12.5Z\" />\n\t\t\t\t</svg>\n\t\t\t</button>\n\n\t\t\t<div class=\"calendar\"\n                 [hidden]=\"!pickerVisible\">\n\t\t\t\t<ngx-time [selectedHour]=\"selectedHour\"\n                          [selectedMinute]=\"selectedMinute\"\n                          [use24HourClock]=\"use24HourClock\"\n                          [doNotCloseOnDateSet]=\"doNotCloseOnDateSet\"\n                          (closeDatePicker)=\"setPickerVisible(!$event)\"\n                          (selectedHourChange)=\"setHourNow($event)\"\n                          (selectedMinuteChange)=\"setMinuteNow($event)\"></ngx-time>\n\t\t\t</div>\n\t\t</div>\n\t</div>\n</div>\n",
            encapsulation: ViewEncapsulation.None,
            providers: [
                {
                    provide: NG_VALUE_ACCESSOR,
                    useExisting: forwardRef(function () { return TimePickerComponent_1; }),
                    multi: true,
                },
            ]
        })
    ], TimePickerComponent);
    return TimePickerComponent;
}());
export { TimePickerComponent };
