// ngx-datetime-picker/esm5/components/date/date.component.js
import * as tslib_1 from "tslib";
import { Component, Input, Output, EventEmitter, ViewChild, ViewEncapsulation } from '@angular/core';
import { DateService } from '../../services/date.service';
var DateComponent = /** @class */ (function () {
    function DateComponent(dateService) {
        this.dateService = dateService;
        this.doNotCloseOnDateSet = false;
        this.min = null;
        this.max = null;
        this.use24HourClock = false;
        this.selectedDateChange = new EventEmitter();
        this.closeDatePicker = new EventEmitter();
        this.alreadySpecifiedMonth = false;
        this.alreadySpecifiedYear = false;
        this.showMonthSelection = false;
        this.showYearSelection = false;
    }
    Object.defineProperty(DateComponent.prototype, "selectedMonth", {
        get: function () {
            if (this.selectedDate == null) {
                return new Date().getMonth() + 1;
            }
            return this.selectedDate.getMonth() + 1;
        },
        set: function (month) {
            var newDate = new Date(this.selectedDate);
            if (month == null) {
                month = new Date().getMonth();
            }
            newDate.setMonth(month - 1);
            this.loadCalendarMonth(newDate);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DateComponent.prototype, "selectedDay", {
        get: function () {
            if (this.selectedDate == null) {
                return new Date().getDate();
            }
            return this.selectedDate.getDate();
        },
        set: function (day) {
            var newDate = new Date(this.selectedDate);
            newDate.setDate(day);
            this.loadCalendarMonth(newDate);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DateComponent.prototype, "selectedYear", {
        get: function () {
            if (this.selectedDate == null) {
                return new Date().getFullYear();
            }
            return this.selectedDate.getFullYear();
        },
        set: function (year) {
            var newDate = new Date(this.selectedDate);
            newDate.setFullYear(year);
            this.loadCalendarMonth(newDate);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DateComponent.prototype, "selectedMonthText", {
        get: function () {
            return this.dateService.getMonthText(this.selectedDate);
        },
        enumerable: true,
        configurable: true
    });
    DateComponent.prototype.setMonth = function (i) {
        this.selectedMonth = i;
        this.showMonthSelection = false;
        this.alreadySpecifiedMonth = true;
        if (!this.alreadySpecifiedYear) {
            this.showYearSelection = true;
        }
    };
    DateComponent.prototype.setSelectedDate = function (date, hour, minutes) {
        if (this.includeTime && !!date && !!this.selectedDate) {
            date.setHours(this.selectedDate.getHours(), this.selectedDate.getMinutes());
        }
        if (!date) {
            date = new Date();
        }
        if (this.min && date < new Date(this.min)) {
            date = new Date(this.min);
        }
        if (this.max && date > new Date(this.max)) {
            date = new Date(this.max);
        }
        //load calendarMonth will set the selected date;
        this.loadCalendarMonth(date);
        if (hour != null) {
            this.selectedDate.setHours(hour);
        }
        if (minutes != null) {
            this.selectedDate.setMinutes(minutes);
        }
        this.selectedDateChange.emit(this.selectedDate);
        this.highlightedDate = this.selectedDate;
        this.selectedHour = date.getHours();
        this.selectedMinute = date.getMinutes();
        if (!this.doNotCloseOnDateSet) {
            this.closePicker();
        }
    };
    DateComponent.prototype.setYear = function (year) {
        this.selectedYear = year;
        this.showYearSelection = false;
        this.alreadySpecifiedYear = true;
        if (!this.alreadySpecifiedMonth) {
            this.showMonthSelection = true;
        }
    };
    DateComponent.prototype.loadCalendarMonth = function (date) {
        if (date == null) {
            date = new Date();
        }
        var shouldReloadCalendar = (this.selectedMonth != (date.getMonth() + 1) || this.selectedYear != date.getFullYear());
        this.selectedDate = date;
        if (shouldReloadCalendar) {
            this.availableDays = tslib_1.__spread(this.dateService.getDateList(this.selectedMonth, this.selectedYear));
        }
    };
    DateComponent.prototype.ngOnInit = function () {
        var _this = this;
        this.months = this.dateService.getMonths();
        this.years = this.dateService.getAvailableYears();
        // subscribing to it's own event emitter to set the selected year position
        this.selectedDateChange.subscribe(function () {
            _this.scrollToMonth();
            _this.scrollToYear();
        });
        //If no date is selected then default to today's date.
        if (!this.selectedDate) {
            if (this.min && new Date(this.min) > new Date()) {
                this.selectedDate = new Date(this.min);
            }
            else if (this.max && new Date(this.max) < new Date()) {
                this.selectedDate = new Date(this.max);
            }
            else {
                this.selectedDate = new Date();
            }
        }
        if (typeof this.selectedDate == 'string') {
            this.selectedDate = new Date(this.selectedDate);
        }
        if (this.includeTime) {
            this.selectedHour = this.selectedDate.getHours();
        }
        if (this.includeTime) {
            this.selectedMinute = this.selectedDate.getMinutes();
        }
        this.highlightedDate = this.selectedDate;
        this.availableDays = tslib_1.__spread(this.dateService.getDateList(this.selectedMonth, this.selectedYear));
    };
    DateComponent.prototype.canSelectYear = function (year) {
        return this.dateService.canSelectYear(year, this.min, this.max);
    };
    DateComponent.prototype.canSelectMonth = function (month) {
        return this.dateService.canSelectMonth(month, this.selectedYear, this.min, this.max);
    };
    DateComponent.prototype.canSelectDay = function (day, month) {
        return this.dateService.canSelectDay(day, month, this.selectedYear, this.min, this.max);
    };
    DateComponent.prototype.scrollToYear = function () {
        var _this = this;
        // setTime out is being used since I need this code to excute next, if not the change won't be visible until the second click
        setTimeout(function () {
            if (_this.yearSelect && _this.yearSelect.nativeElement) {
                var selectContainer = _this.yearSelect.nativeElement;
                var selectedYear = selectContainer.querySelector('.calendar--year__selected');
                selectContainer.scrollTop = selectedYear.offsetTop - (selectContainer.clientHeight / 2) - (selectedYear.clientHeight);
            }
        });
    };
    DateComponent.prototype.scrollToMonth = function () {
        var _this = this;
        // setTime out is being used since I need this code to execute next, if not the change won't be visible until the second click
        setTimeout(function () {
            if (_this.monthSelect && _this.monthSelect.nativeElement) {
                var selectContainer = _this.monthSelect.nativeElement;
                var selectedMonth = selectContainer.querySelector('.calendar--month__selected');
                selectContainer.scrollTop = selectedMonth.offsetTop - (selectContainer.clientHeight / 2) - (selectedMonth.clientHeight);
            }
        });
    };
    DateComponent.prototype.previousMonth = function () {
        this.alreadySpecifiedMonth = false;
        var previousMonth = new Date(this.selectedDate);
        //because javascript sets months based on a 0 index need to jump back 2 to go to the previous month.
        previousMonth.setMonth(this.selectedMonth - 2);
        this.loadCalendarMonth(previousMonth);
    };
    DateComponent.prototype.nextMonth = function () {
        this.alreadySpecifiedMonth = false;
        var nextMonth = new Date(this.selectedDate);
        /// same as above but since selected month is 1-12 the index is already the next month.
        nextMonth.setMonth(this.selectedMonth);
        this.loadCalendarMonth(nextMonth);
    };
    DateComponent.prototype.toggleMonthMenu = function () {
        this.scrollToMonth();
        this.showMonthSelection = !this.showMonthSelection;
    };
    DateComponent.prototype.toggleYearMenu = function () {
        this.scrollToYear();
        this.showYearSelection = !this.showYearSelection;
    };
    DateComponent.prototype.closePicker = function () {
        this.alreadySpecifiedMonth = false;
        this.alreadySpecifiedYear = false;
        this.closeDatePicker.emit(false);
    };
    DateComponent.ctorParameters = function () { return [
        { type: DateService }
    ]; };
    tslib_1.__decorate([
        Input()
    ], DateComponent.prototype, "selectedDate", void 0);
    tslib_1.__decorate([
        Input()
    ], DateComponent.prototype, "includeTime", void 0);
    tslib_1.__decorate([
        Input()
    ], DateComponent.prototype, "doNotCloseOnDateSet", void 0);
    tslib_1.__decorate([
        Input()
    ], DateComponent.prototype, "min", void 0);
    tslib_1.__decorate([
        Input()
    ], DateComponent.prototype, "max", void 0);
    tslib_1.__decorate([
        Input()
    ], DateComponent.prototype, "use24HourClock", void 0);
    tslib_1.__decorate([
        Output()
    ], DateComponent.prototype, "selectedDateChange", void 0);
    tslib_1.__decorate([
        Output()
    ], DateComponent.prototype, "closeDatePicker", void 0);
    tslib_1.__decorate([
        ViewChild('yearSelect')
    ], DateComponent.prototype, "yearSelect", void 0);
    tslib_1.__decorate([
        ViewChild('monthSelect')
    ], DateComponent.prototype, "monthSelect", void 0);
    DateComponent = tslib_1.__decorate([
        Component({
            selector: 'ngx-date',
            template: "<div class=\"calendar\">\n\t<button type=\"button\"\n            class=\"ngx-picker--btn ngx-picker--btn__month\"\n            (click)=\"toggleMonthMenu(); showYearSelection = false\">\n        {{ selectedMonthText }}\n    </button>\n\n\t<button type=\"button\"\n            class=\"ngx-picker--btn ngx-picker--btn__year\"\n            (click)=\"toggleYearMenu(); showMonthSelection = false\">\n        {{ selectedYear }}\n    </button>\n\n\t<span class=\"calendar--previous-and-next\">\n\t\t<button type=\"button\"\n                class=\"ngx-picker--btn ngx-picker--btn__previous\"\n\t\t\t(click)=\"previousMonth()\">\n            &lt;\n        </button>\n\n\t\t<button type=\"button\"\n                class=\"ngx-picker--btn ngx-picker--btn__next\"\n                (click)=\"nextMonth()\">\n            &gt;\n        </button>\n\t</span>\n\n\t<div class=\"calendar--months-select\"\n         #monthSelect\n         [hidden]=\"!showMonthSelection\">\n\t\t<span *ngFor=\"let month of months; let i = index ;\"\n              class=\"calendar--month\"\n\t\t\t  [class.calendar--month__selected]=\"i + 1 == selectedMonth\"\n\t\t\t  [class.calendar--month__disabled]=\"!canSelectMonth(i)\"\n\t\t\t  (click)=\"setMonth(i + 1)\">\n            {{ month }}\n        </span>\n\t</div>\n\n\t<div class=\"calendar--years-select\"\n         #yearSelect\n         [hidden]=\"!showYearSelection\">\n\t\t<span *ngFor=\"let year of years\"\n              class=\"calendar--year\"\n\t\t\t  [class.calendar--year__selected]=\"selectedYear == year\"\n\t\t\t  [class.calendar--year__disabled]=\"!canSelectYear(year)\"\n\t\t\t  (click)=\"this.setYear(year)\">\n            {{ year }}\n\t\t</span>\n\t</div>\n\n\t<div *ngIf=\"showMonthSelection == false && showYearSelection == false\"\n         class=\"calendar--days-and-weeks\">\n        <div class=\"calendar--days-of-week\">\n            <span class=\"calendar--day-of-week\">Su</span>\n            <span class=\"calendar--day-of-week\">Mo</span>\n            <span class=\"calendar--day-of-week\">Tu</span>\n            <span class=\"calendar--day-of-week\">We</span>\n            <span class=\"calendar--day-of-week\">Th</span>\n            <span class=\"calendar--day-of-week\">Fr</span>\n            <span class=\"calendar--day-of-week\">Sa</span>\n        </div>\n\n        <div class=\"calendar--days-select\">\n            <span *ngFor=\"let day of availableDays\"\n                  class=\"calendar--day\"\n                  [class.calendar--day__muted]=\"day.month != selectedMonth\"\n                  [class.calendar--day__selected]=\"(\n                      day.date.getDate() == highlightedDate?.getDate()\n                      && day.date.getFullYear() == highlightedDate?.getFullYear()\n                      && day.date.getMonth() == highlightedDate?.getMonth()\n                  )\"\n                  [class.calendar--day__disabled]=\"!canSelectDay(day.day, day.date.getMonth())\"\n                  (click)=\"setSelectedDate(day.date)\">\n                {{ day.day}}\n            </span>\n        </div>\n\t</div>\n\n\t<ngx-time *ngIf=\"includeTime\"\n\t\t\t  [selectedHour]=\"selectedHour\"\n\t\t\t  [selectedMinute]=\"selectedMinute\"\n\t\t\t  (selectedHourChange)=\"setSelectedDate(highlightedDate,$event)\"\n\t\t\t  (selectedMinuteChange)=\"setSelectedDate(highlightedDate,null, $event)\"\n              (closeDatePicker)=\"closePicker()\"\n\t\t\t  [use24HourClock]=\"use24HourClock\"></ngx-time>\n</div>\n",
            encapsulation: ViewEncapsulation.None
        })
    ], DateComponent);
    return DateComponent;
}());
export { DateComponent };
